lib/recursive-methods.rb

```diff
@@ -1,49 +1,104 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+	return 1 if n == 0
+	raise ArgumentError, "Sorry! I can't calculate negative numbers!" if n < 0
+	return (n * factorial(n - 1))
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n2)
 def reverse(s)
```

Comment:
I assume you meant n^2 for space complexity. However since you're creating a new array by copying n-1 elements over into the new array... this is O(n^2) for time complexity too.